main/acle.md

```diff
@@ -2696,10 +2701,12 @@ following:
   in one of the translation units.
   * Implicitly, without this attribute,
   * or explicitly providing the `default` in the attribute.
-* All instances of the versions shall share the same function
-  signature and calling convention.
```

Comment:
I agree the wording should be improved, but I believe there is a need for an adjustment here.
This was specifically something requested by upstream C maintainers in GCC.

The justification is that we likely don't want to require exactly the same signatures, as it
is common for declarations and implementations to differ in signature while still being compatible:

Something like
```C++
int __attribute__((target_version("default"))) fn (int x[]);

int __attribute__((target_version("default"))) fn (int x[5]) {
    return 1;
}
```

Or:
```C++
int __attribute__((target_version("default"))) fn (bool x = true);

int __attribute__((target_version("default"))) fn (bool x) {
    return 1;
}
```
For a more CPP example.

(LLVM is happy with both these currently)

But if this is allowed at some level then we need to define if information for one version should be shared with others, eg a case like:

```C++
int __attribute__((target_version("default"))) fn (bool c = true);
int __attribute__((target_version("sve"))) fn (bool c);

int __attribute__((target_version("default"))) fn (bool c) {
    return 1;
}

int bar() {
    fn();
}
```
(Again allowed by LLVM)
Should this default to true for the SVE version? I would assume so but its not stipulated?

We could disallow it completely, but that seems overly restrictive and counter to C and C++ intuitions.

Reply:
Oh wow, okay. That example gives me really bad mojo, and I would advocate we should disallow that if possible, though I wont die on that hill. 
I can live with the default signature dominating, but request we document that in the ACLE.

I would advocate for documenting it in the ACLE. I think the purpose of the ACLE is to document this sort of edge case so we can ensure matched behavior between compilers, and this point came up because my natural inclination when implementing it in GCC was to go another direction. 

This was specifically requested by the GCC C front end maintainer for what I imagine was very much this sort of reason.

Reply:
Ah okay I see, thank you. I have changed the patch to document this rule a little clearer.

As this came up with the GCC front end maintainers I would prefer to clarify this case.

Reply:
The wording seems better. Perhaps it is worth adding the example https://godbolt.org/z/Enn15Y93e without the codegen but explaining that the value 2 will be passed when calling.

Reply:
Just to check, do you mean adding this directly to the ACLE?

Reply:
Added some now :)